COBOL-TEST-RUNNER: Revised bin/run.sh.

borderite opened the pull request:
- The revised version of bin/run.sh examines if the last four lines of the output from the test program show the summary of the test results. Because absence of the summary implies that the program prematurely teminated skipping some or all tests,, bin/run.sh fails the solution.

- To verify that the modified test runner behaves in the intended way, one example case "example-premature-termination" has been added to the tests directory.

- The docker build command seems to fail without a "--netowrk host" flag. The flag has been added to the docker build command in bin/run-in-docker.sh and bin/run-tests-in-docker.sh.

borderite committed:
COBOL-TEST-RUNNER: Revised bin/run.sh.

- The revised version of bin/run.sh examines if the last four lines of
the output from the test program show the summary of the test
results. Because absence of the summary implies that the program
prematurely teminated skipping some or all tests,, bin/run.sh fails
the solution.

- To verify that the modified test runner behaves in the intended way,
one example case "example-premature-termination" has been added to the
tests directory.

- The docker build command seems to fail without a "--netowrk host"
flag. The flag has been added to the docker build command in
bin/run-in-docker.sh and bin/run-tests-in-docker.sh.

IsaacG commented:
Reopened for discussion. Maintainers, feel free to close this out if that's preferred.

borderite committed:
COBOL-TEST-RUNNER: Removed backup files created by Emacs.

- Also, added "*~" to .gitignore.

glennj commented on bin/run.sh:
```diff
@@ -39,7 +39,13 @@ test_output=$(bash ${solution_dir}/test.sh ${slug} 2>&1)
 # just executed that tested the implementation file
 if [ $? -eq 0 ]; then
-    jq -n '{version: 1, status: "pass"}' > ${results_file}
+    if echo "$test_output" | tail -4 | grep -Pz '^ *[[:digit:]]+ TEST CASES WERE EXECUTED\n *[[:digit:]]+ PASSED\n *[[:digit:]]+ FAILED\n={49}' 2>&1 1>/dev/null; then
```
Instead of `grep ... 2>&1 1>/dev/null` (which is subtly incorrect), you can do `grep -q ...` which outputs nothing but sets the exit status to the expected value.

The current code:
* redirects file descriptor 2 to what fd 1 is currently using (/dev/stdout)
* redirects fd 1 to /dev/null

The problem is that fd 2 is still using /dev/stdout.

borderite replied:
Oops, I meant to write  `>/ dev/null 2>&1`. Yes, use of `-q` is great.

borderite replied:
I have made the suggested change.

kotp commented on .gitignore:
```diff
@@ -2,4 +2,5 @@ tests/*/results.json
-tests/*/test.cob
\ No newline at end of file
+tests/*/test.cob
+*~
```
This entry does not belong as it has nothing to do with the project.

Also, the last line is missing an `EOL`.

borderite replied:
Done.

kotp reviewed (changes requested):
User choice items should go in the Users excludes file, as we should not have to maintain all and any user choice items.  Keep the project `.gitignore` for the project artifact that it creates but should not be included in the repository.

kotp commented on .gitignore:
```diff
@@ -2,4 +2,5 @@ tests/*/results.json
 tests/*/test
-tests/*/test.cob
\ No newline at end of file
+tests/*/test.cob
+*~
```
```suggestion
tests/*/test.cob

```

IsaacG commented on bin/run.sh:
```diff
@@ -39,7 +39,13 @@ test_output=$(bash ${solution_dir}/test.sh ${slug} 2>&1)
 if [ $? -eq 0 ]; then
-    jq -n '{version: 1, status: "pass"}' > ${results_file}
+    if echo "$test_output" | tail -4 | grep -Pz '^ *[[:digit:]]+ TEST CASES WERE EXECUTED\n *[[:digit:]]+ PASSED\n *[[:digit:]]+ FAILED\n={49}' 2>&1 1>/dev/null; then
+        jq -n '{version: 1, status: "pass"}' > ${results_file}
```
```suggestion
        jq -n '{version: 1, status: "pass"}' > "${results_file}"
```

borderite committed:
COBOL-TEST-RUNNER: Removed *~ from .gitignore per kotp's suggestion.

IsaacG commented on bin/run.sh:
```diff
@@ -39,7 +39,13 @@ test_output=$(bash ${solution_dir}/test.sh ${slug} 2>&1)
+    if echo "$test_output" | tail -4 | grep -Pz '^ *[[:digit:]]+ TEST CASES WERE EXECUTED\n *[[:digit:]]+ PASSED\n *[[:digit:]]+ FAILED\n={49}' 2>&1 1>/dev/null; then
+        jq -n '{version: 1, status: "pass"}' > ${results_file}
+    else
+        sanitized_test_output=$(printf "${test_output}" | sed '1,/^COMPILE AND RUN TEST$/d' | sed '/warning: ignoring redundant \. \[-Wothers\]/ d' | sed '/test.cob: in paragraph .\(UT-BEFORE-EACH\|UT-AFTER-EACH\|UT-LOOKUP-FILE\|UT-BEFORE\)./ d' )
+        sanitized_test_output="${sanitized_test_outputtest} $(printf " \nSOME TESTS WERE NOT PERFORMED. A PARAGRAPH BEING TESTED MUST HAVE FORCED THE PROGRAM EXECUTION TO TERMINATE (E.G. USING STOP RUN, GOBACK, ETC.)")"
+        jq -n --arg output "${sanitized_test_output}" '{version: 1, status: "fail", message: $output}' > ${results_file}
```
```suggestion
        sanitized_test_output=$(printf "%s\n" "${test_output}" | sed '1,/^COMPILE AND RUN TEST$/d; /warning: ignoring redundant \. \[-Wothers\]/ d;  /test.cob: in paragraph .\(UT-BEFORE-EACH\|UT-AFTER-EACH\|UT-LOOKUP-FILE\|UT-BEFORE\)./ d' )
        printf -v sanitized_test_output "%s\nSOME TESTS WERE NOT PERFORMED. A PARAGRAPH BEING TESTED MUST HAVE FORCED THE PROGRAM EXECUTION TO TERMINATE (E.G. USING STOP RUN, GOBACK, ETC.)" "${sanitized_test_outputtest}"
        jq -n --arg output "${sanitized_test_output}" '{version: 1, status: "fail", message: $output}' > "${results_file}"
```

bash printf supports `-v`

IsaacG commented on bin/run.sh:
```diff
@@ -49,6 +55,8 @@ else
         sanitized_test_output=$(printf "${test_output}" | sed '1,/^COMPILE AND RUN TEST$/d' | sed '/warning: ignoring redundant \. \[-Wothers\]/ d' | sed '/test.cob: in paragraph .\(UT-BEFORE-EACH\|UT-AFTER-EACH\|UT-LOOKUP-FILE\|UT-BEFORE\)./ d' )
     else
         sanitized_test_output="${test_output} $(printf " \nSOMETHING WENT WRONG DURING TEST SETUP, PLEASE OPEN A TICKET AT: https://github.com/exercism/cobol/issues/new")"
+    jq -n --arg output "${sanitized_test_output}" '{version: 1, status: "fail", message: $output}' > ${results_file}
+        
```
```suggestion
        printf -v sanitized_test_output "%s\nSOMETHING WENT WRONG DURING TEST SETUP, PLEASE OPEN A TICKET AT: https://github.com/exercism/cobol/issues/new" "${test_output}"
        jq -n --arg output "${sanitized_test_output}" '{version: 1, status: "fail", message: $output}' > "${results_file}"
```

IsaacG commented on tests/example-premature-termination/test.sh:
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
```
If this relies on bash, can all shell scripts be upgraded from sh to bash?

borderite replied:
The test.sh in the tests directory is essentially the same as the file with the same name included in every exercise in the COBOL track. If we need to change it, it should be changed in the COBOL exercise repository first.

IsaacG replied:
The shebang controls the interpreter used. The shebang and file name on non-Windows systems are entirely independent.

borderite replied:
My reponse was probably somewhat misleading. The shebang line of test.sh is correct, as test.sh is using bash-specific feature in it. Does it have to be in this way? Maybe not. But all test.sh files in the track currently require bash. (For windows platforms, the COBOL track provides powershell script, test.ps1.)

Because bash is available in the test-runner environment, it is possible to upgrade all scripts in the runner to bash. But I don't see why you might want to do so.

IsaacG replied:
If bash is the shell installed in the environment, odds are high that `/bin/sh` is actually running bash in POSIX mode. Using `/bin/bash` ie running bash in bash mode allows you to use bash features in the other scripts. The additional features are faster, safer and more flexible. IIUC removing the POSIX checks also makes things faster as fewer checks are run. If there's a reason to prefer POSIX sh, then it makes sense to use POSIX sh. If there is no reason to prefer POSIX sh, then there's no reason not to switch them to bash.

borderite replied:
Now I see the picture. I agree that it is a good idea to replace sh with bash. I have made the changes.

IsaacG commented on tests/example-premature-termination/test.sh:
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )/
+SLUG=$(basename "${SCRIPT_DIR}")
```
```suggestion
full_path=$(readlink -e "${BASH_SOURCE[0]}" )
script_dir=${full_path%/*}
slug=${script_dir##*/}
```

IsaacG commented on tests/example-premature-termination/test.sh:
```diff
@@ -0,0 +1,20 @@
+SLUG=$(basename "${SCRIPT_DIR}")
+COBOLCHECK=./bin/cobolcheck
+
+WHICH_COBOLCHECK=$(which cobolcheck)
+if [[ $? -eq 0 ]] ; then
+    echo "Found cobolcheck, using $COBOLCHECK"
+    COBOLCHECK=$WHICH_COBOLCHECK
+elif [ ! -f $SCRIPT_DIR/bin/cobolcheck ]; then
+    echo "Cobolcheck not found, try to fetch it."
+    cd $SCRIPT_DIR/bin/
+    ./fetch-cobolcheck
+fi
+cd $SCRIPT_DIR
+$COBOLCHECK -p $SLUG
```
```suggestion
if cobolcheck_type=$(type cobolcheck); then
    echo "Found cobolcheck, ${cobolcheck_type}"
    COBOLCHECK=cobolcheck
elif [[ ! -x $SCRIPT_DIR/bin/cobolcheck ]; then
    echo "cobolcheck not found, try to fetch it."
    ./bin/fetch-cobolcheck
fi
"${COBOLCHECK}" -p "${slug}"
```

borderite replied:
@IsaacG  Though I have noticed lots of style problems while going over the script, I decided to focus on fixing the particular issue discussed in the forum.  Of course, I like the changes you suggested. But I am wondering if we should make the changes in this PR.

IsaacG replied:
It can definitely be split into different PRs.

borderite replied:
I support the split!

IsaacG replied:
Any new lines added in this PR should follow best shell practices, though!

borderite replied:
Sure. I've incorporated the changes you suggested within the limit.

borderite committed:
COBOL-TEST-RUNNER: Corrected an error in use of grep.

- Removed a wrong file direction in using the grep command and add -q
switch to surpress the output from it.

borderite committed:
COBOL-TEST-RUNNER: Modify the shebang lines of the shell scripts.

- Replaced sh with bash in the shebang lines of the four scripts, per
Isaac's suggestion.

IsaacG commented on bin/run.sh:
```diff
@@ -39,7 +39,13 @@ test_output=$(bash ${solution_dir}/test.sh ${slug} 2>&1)
 # Write the results.json file based on the exit code of the command that was 
 # just executed that tested the implementation file
 if [ $? -eq 0 ]; then
```
If you're using back, you can use `if (( $? == 0 ))` ... but that can also be another PR.

borderite replied:
I think that it should be another PR.

IsaacG commented on bin/run.sh:
```diff
@@ -49,6 +55,8 @@ else
         sanitized_test_output=$(printf "${test_output}" | sed '1,/^COMPILE AND RUN TEST$/d' | sed '/warning: ignoring redundant \. \[-Wothers\]/ d' | sed '/test.cob: in paragraph .\(UT-BEFORE-EACH\|UT-AFTER-EACH\|UT-LOOKUP-FILE\|UT-BEFORE\)./ d' )
     else
         sanitized_test_output="${test_output} $(printf " \nSOMETHING WENT WRONG DURING TEST SETUP, PLEASE OPEN A TICKET AT: https://github.com/exercism/cobol/issues/new")"
```
```suggestion
        printf -v "%s\nSOMETHING WENT WRONG DURING TEST SETUP, PLEASE OPEN A TICKET AT: https://github.com/exercism/cobol/issues/new" "${test_output}"
```

glennj replied:
@IsaacG you're missing the variable name there:
```sh
printf -v sanitized_test_output "%s\nSOMETHING WENT WRONG DURING TEST SETUP, PLEASE OPEN A TICKET AT: https://github.com/exercism/cobol/issues/new" "${test_output}"
```

glennj replied:
@borderite are you sure you want people opening GitHub issues for the Cobol repo? Who is promising to monitor that? I would suggest you direct people to the forum: https://forum.exercism.org/c/programming/cobol/75

borderite replied:
@glennj  A good point. Because this PR strictly focuses on  detection of premature termination,
 it is beyond the scope of the current PR. Please feel to start the discussion in the forum.

kotp commented on tests/example-premature-termination/tst/example-premature-termination/example-premature-termination.cut:
```diff
@@ -0,0 +1,5 @@
+
+TestCase "sucess 1"
+        PERFORM DO-SOMETHING
+	EXPECT WS-RESULT = "TRUE"
```
```suggestion
	EXPECT WS-RESULT = "TRUE"
	
```

EOL

kotp replied:
Also, commit messages have mixed tense.  Commits do something in the moment, rather than in the past.

Not necessarily a "need to fix" but just be aware.

Might review Chris Beams' article[1].

[1]: https://cbea.ms/git-commit/

borderite committed:
Update tests/example-premature-termination/tst/example-premature-termination/example-premature-termination.cut

Convert a tab to spaces.

Co-authored-by: Victor Goff <keeperotphones@gmail.com>

kotp commented on tests/example-premature-termination/tst/example-premature-termination/example-premature-termination.cut:
```diff
@@ -2,4 +2,5 @@ TESTSUITE 'Testsuite example-premature-termination'
-	EXPECT WS-RESULT = "TRUE"
\ No newline at end of file
+	EXPECT WS-RESULT = "TRUE"
+	
```
```suggestion
	EXPECT WS-RESULT = "TRUE"

```

Trailing white space with missing EOL is not better.

This should fix that.

borderite committed:
Update tests/example-premature-termination/tst/example-premature-termination/example-premature-termination.cut

Co-authored-by: Victor Goff <keeperotphones@gmail.com>

borderite committed:
Change styles in a few lines.

- Replace the combination of command substituion and printf with
printf -v in bin/run.sh.

- Quote a couple of shell variables in bin/run.sh

- Added a missing new line in the cut file of the
example-premature-termination test.

- Reflecting the changes in run.sh, adjusted expected_results.json in
the example-premature-termination test.

IsaacG commented on bin/run.sh:
```diff
@@ -49,6 +55,8 @@ else
         sanitized_test_output=$(printf "${test_output}" | sed '1,/^COMPILE AND RUN TEST$/d' | sed '/warning: ignoring redundant \. \[-Wothers\]/ d' | sed '/test.cob: in paragraph .\(UT-BEFORE-EACH\|UT-AFTER-EACH\|UT-LOOKUP-FILE\|UT-BEFORE\)./ d' )
     else
         sanitized_test_output="${test_output} $(printf " \nSOMETHING WENT WRONG DURING TEST SETUP, PLEASE OPEN A TICKET AT: https://github.com/exercism/cobol/issues/new")"
+    jq -n --arg output "${sanitized_test_output}" '{version: 1, status: "fail", message: $output}' > ${results_file}
```
```suggestion
    jq -n --arg output "${sanitized_test_output}" '{version: 1, status: "fail", message: $output}' > "${results_file}"
```

IsaacG committed:
Update bin/run.sh

borderite committed:
Modify test.sh following the best practice as suggested by Isaac.

IsaacG commented on tests/example-premature-termination/test.sh:
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+full_path=$(readlink -e "${BASH_SOURCE[0]}" )
+script_dir=${full_path%/*}
+slug=${script_dir##*/}
+
+if cobolcheck_type=$(type cobolcheck); then
+    echo "Found cobolcheck, ${cobolcheck_type}"
+    COBOLCHECK=cobolcheck
+elif [[ ! -x $SCRIPT_DIR/bin/cobolcheck ]]; then
+    echo "cobolcheck not found, try to fetch it."
+    ./bin/fetch-cobolcheck
+fi
+
+cd $script_dir
+"${COBOLCHECK}" -p "${slug}"
```
In the `else` case, `COBOLCHECK` doesn't get set. Should it use `$SCRIPT_DIR/bin/cobolcheck`?

```suggestion
if cobolcheck_type=$(type cobolcheck); then
    echo "Found cobolcheck, ${cobolcheck_type}"
    COBOLCHECK=cobolcheck
else
    COBOLCHECK=$SCRIPT_DIR/bin/cobolcheck
fi
if [[ ! -x $COBOLCHECK ]]; then
    echo "cobolcheck not found, try to fetch it."
    ./bin/fetch-cobolcheck
fi

cd "${script_dir}"
"${COBOLCHECK}" -p "${slug}"
```

borderite replied:
@IsaacG  Your suggested change for fixing undefined COBOLCHECK has been incorporated.

borderite committed:
Modify test.sh to make a correction aused by the previous change.

- The previous revision removed
     COBOLCHECK=./bin/cobolcheck
  near the top of the file. This causes a problem in a certain case.
- The problem has been fixed following Isaac's suggestion.

IsaacG commented on tests/example-premature-termination/test.sh:
```diff
@@ -0,0 +1,21 @@
+    COBOLCHECK=$SCRIPT_DIR/bin/cobolcheck
+fi
+if [[ ! -x $COBOLCHECK ]]; then
+    echo "cobolcheck not found, try to fetch it."
+    ./bin/fetch-cobolcheck
```
Do you want to check that this succeeds?

borderite replied:
Before commiting changes, I run bin/run-tests-in-docker.sh. Of course, the current version passes the test.

IsaacG replied:
Do you check both the `if` and `else` branch? Is it possible that `fetch-cobolcheck` can ever fail?
